File: neml2_models/src/PK2HydrostaticStress.cxx
// Copyright 2024, UChicago Argonne, LLC
// All Rights Reserved
// Software Name: NEML2 -- the New Engineering material Model Library, version 2
// By: Argonne National Laboratory
// OPEN SOURCE LICENSE (MIT)
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

#include "PK2HydrostaticStress.h"
#include <neml2/base/Registry.h>

#include <neml2/tensors/Scalar.h>
#include <neml2/tensors/R2.h>
#include <neml2/tensors/functions/einsum.h>
#include <neml2/tensors/functions/symmetrization.h>
#include <neml2/tensors/functions/inv.h>
#include <neml2/tensors/functions/det.h>

namespace neml2
{
register_NEML2_object(PK2HydrostaticStress);

OptionSet
PK2HydrostaticStress::expected_options()
{
  OptionSet options = Model::expected_options();
  options.doc() =
      "Phase-change PK2 stress contribution, "
      "\\f$ \\mathbf{S}^{pc} = -J \\, \\sigma_h \\, \\mathbf{F}^{-1}\\mathbf{F}^{-T} "
      "= -J \\, \\sigma_h \\, \\mathbf{C}^{-1} \\f$, "
      "with \\f$ J = \\det(\\mathbf{F}) \\f$.";

  options.set_input("hydrostatic_stress") = VariableName(STATE, "sigma_h");
  options.set("hydrostatic_stress").doc() = "Hydrostatic Cauchy stress offset";

  options.set_input("deformation_gradient") = VariableName(FORCES, "F");
  options.set("deformation_gradient").doc() = "Deformation gradient";

  options.set_output("pk2_stress") = VariableName(STATE, "Spc");
  options.set("pk2_stress").doc() = "Phase-change PK2 stress contribution";

  return options;
}

PK2HydrostaticStress::PK2HydrostaticStress(const OptionSet & options)
  : Model(options),
    _Spc(declare_output_variable<R2>("pk2_stress")),
    _sh(declare_input_variable<Scalar>("hydrostatic_stress")),
    _F(declare_input_variable<R2>("deformation_gradient"))
{
}

void
PK2HydrostaticStress::set_value(bool out, bool dout_din, bool /*d2out_din2*/)
{
  const auto F = _F();
  const auto Finv = neml2::inv(F);
  const auto Cinv = Finv * Finv.transpose();
  const auto J = neml2::det(F);
  const auto a = J * _sh;

  if (out)
    _Spc = -a * Cinv;

  if (dout_din)
  {
    _Spc.d(_sh) = -J * Cinv;

    const auto term1 = einsum("...im,...nj->...ijmn", {Finv, Cinv});
    const auto term2 = einsum("...in,...jm->...ijmn", {Cinv, Finv});
    const auto term3 = einsum("...nm,...ij->...ijmn", {Finv, Cinv});

    _Spc.d(_F) = a * (term1 + term2 - term3);
  }
}

} // namespace neml2

File: neml2_models/python/PK2HydrostaticStress.py
from __future__ import annotations

from neml2.factory import register_neml2_object
from neml2.models.chain_rule import ChainRuleDict
from neml2.models.model import Model
from neml2.schema import HitSchema, input, output
from neml2.types import R2, Scalar, det, inv, sym, tr


@register_neml2_object("PK2HydrostaticStress")
class PK2HydrostaticStress(Model):
    r"""Phase-change PK2 stress contribution
    $\mathbf{S}^{pc} = -J\,\sigma_h\,\mathbf{F}^{-1}\mathbf{F}^{-T}$, $J = \det\mathbf{F}$.
    """

    hit = HitSchema(
        input("hydrostatic_stress", Scalar, "Hydrostatic Cauchy stress offset"),
        input("deformation_gradient", R2, "Deformation gradient"),
        output("pk2_stress", R2, "Phase-change PK2 stress contribution"),
    )

    def forward(  # type: ignore[override]
        self,
        hydrostatic_stress: Scalar,
        deformation_gradient: R2,
        *promoted_params,
        v: ChainRuleDict | None = None,
    ):
        sh = hydrostatic_stress
        F = deformation_gradient
        Finv = inv(F)
        Cinv = Finv @ Finv.base.transpose(-2, -1)
        J = det(F)
        a = J * sh
        Spc = Cinv * (-a)
        if v is None:
            return Spc

        def sh_action(V: Scalar) -> R2:
            return Cinv * (-J * V)

        # dS/dF contracted with a tangent V, with G = Finv @ V:
        #   dS(V) = a (G Cinv + Cinv G^T - tr(G) Cinv)
        def F_action(V: R2) -> R2:
            G = Finv @ V
            return (G @ Cinv + Cinv @ G.base.transpose(-2, -1) - Cinv * tr(sym(G))) * a

        return Spc, self.apply_chain_rule(
            v,
            "pk2_stress",
            {"hydrostatic_stress": sh_action, "deformation_gradient": F_action},
            output=Spc,
        )
